[FABN-1572] fabricCaClient.js hygiene (p1)

davidkhala opened the pull request:
- avoid to use self, `this` is clear for arrow function style
- mark as `async` for method has potential to return a promise.

Signed-off-by: david <david-khala@hotmail.com>

davidkhala committed:
[FABN-1572] fabricCaClient.js hygiene (p1)

- avoid to use self, `this` is clear for arrow function style
- mark as `async` for method has potential to return a promise.
- remove ts.d: DiscoveryService.setDiscoverer is not implemented
	- use setTargets instead
- merge duplicated http request event handler codes
- reserve some request error information in new Error

Signed-off-by: david <david-khala@hotmail.com>

bestbeforetoday commented:
Is this ready for review or are you still making changes?

davidkhala commented:
> Is this ready for review or are you still making changes?

Ready for review. 
There is something I would like to change, I will collect them in another 'Part 2' PR.

bestbeforetoday commented on fabric-ca-client/test/FabricCAClient.js:
```diff
@@ -1080,19 +1084,8 @@ describe('FabricCAClient', () => {
 			sinon.assert.calledOnce(requestStub);
 
-			// should call with known
-			const expected = {
```
The test says:
> should call POST with the correct method, request, and signing identity

The test for these values has been removed so it is no longer testing what it is supposed to be testing

davidkhala replied:
@bestbeforetoday Do you mean I should add back a green test case?

bestbeforetoday reviewed (changes requested):
Just a unit test concern

bestbeforetoday commented on fabric-ca-client/test/FabricCAClient.js:
```diff
@@ -1199,33 +1208,9 @@ describe('FabricCAClient', () => {
+			await client.enroll('enough', 'parameters', 'for_this_function_call').should.be.rejectedWith(/Could not parse enroll response/);
 		});
 
-		it('should reject on end point error', async () => {
```
What is the reason for removing this test? In general I want to see all the existing tests continue to pass on a refactor to ensure it does not regress behaviour

davidkhala replied:
@bestbeforetoday 
Hi Mark, 
This test for a corner case which would only happens in rewire test instead of real case. 
As far as I know. Please clarify if I am wrong. 
This test assume multiple error handler exist both in or out of response callback, which I see it is redundant.

davidkhala replied:
But thx, I find that we could register a error handler on response, thus we could handle this case.

bestbeforetoday commented on fabric-ca-client/lib/FabricCAClient.js:
```diff
@@ -70,21 +65,21 @@ const FabricCAClient = class {
 		this._cryptoPrimitives = cryptoPrimitives;
 
-		logger.debug('Successfully constructed Fabric CA client from options - %j', connect_opts);
+		logger.debug(`Successfully constructed Fabric CA client from options - ${JSON.stringify(connect_opts)}`);
```
If somehow `connect_ops` has a circular reference, `JSON.stringify()` will throw an error. In the previous version the logging framework would be using the Node _util_ package, which handles circular references without an error.

Actually both `util.format()` with `%o` and `util.inspect()` do a better job in this case, but not throwing an error in logging is critical.

```
> const a = {};
> const b = {a};
> a.b = b;
> JSON.stringify(a);
Thrown:
TypeError: Converting circular structure to JSON
    at JSON.stringify (<anonymous>)
> util.format('%j', a);
'[Circular]'
> util.format('%o', a);
'{ b: { a: [Circular] } }'
> util.inspect(a);
'{ b: { a: [Circular] } }'
```

davidkhala replied:
Thanks for your clarify and reminds! I would use the code fashion combine with string literal and `util.inspect`

bestbeforetoday commented on fabric-ca-client/lib/FabricCAClient.js:
```diff
@@ -278,75 +246,82 @@ const FabricCAClient = class {
+					if (responseObj.success) {
+						return resolve(responseObj);
+					} else {
+						const err = Error(`fabric-ca request ${api_method} failed with errors [${JSON.stringify(responseObj && responseObj.errors ? responseObj.errors : responseObj)}]`);
```
Again, I would avoid JSON.stringify(). I've been burned by this one before.

bestbeforetoday commented on fabric-ca-client/test/FabricCAClient.js:
```diff
@@ -1286,36 +1293,6 @@ describe('FabricCAClient', () => {
 			testCRL.should.be.equal('bazinga');
 		});
 
-		it('should reject if POST does not return success in the response body', async () => {
```
Unit test removed?

davidkhala replied:
My fault.

bestbeforetoday commented on fabric-ca-client/test/FabricCAClient.js:
```diff
@@ -1286,36 +1293,6 @@ describe('FabricCAClient', () => {
-			await client.generateCRL('revokedBefore', 'revokedAfter', 'expireBefore', 'expireAfter', 'signingIdentity').should.be.rejectedWith('test_fail');
-		});
-
-		it('should reject if POST does not return result in the repsonse body', async () => {
```
Unit test removed?

bestbeforetoday reviewed (changes requested):
Some comments below. Might be easier to push an additional commit so I can more easily see what's changed since the previous one; I will squash the commits at merge time anyway so having multiple commits in the PR is fine

davidkhala commented on fabric-common/lib/impl/bccsp_pkcs11.js:
```diff
@@ -599,8 +599,7 @@ class CryptoSuite_PKCS11 extends CryptoSuite {
 			const rns = new ecsig(signature, 'hex');
-			logger.debug(__func() + 'ECDSA R+S signature: ' +
-				util.inspect(rns, {depth: null}));
+			logger.debug(__func() + 'ECDSA R+S signature: ', rns);
```
This code change aim to solve deprecation warning
`(node:6461) [DEP0079] DeprecationWarning: Custom inspection function on Objects via .inspect() is deprecated`

davidkhala committed:
SHOULD be squashed
- add back the deleted unit test
- avoid to use JSON.stringify to directly user input object in logging code
- more jsDoc cleanup

Signed-off-by: david <david-khala@hotmail.com>

bestbeforetoday commented:
> I consider if we could rearrange them in orderer of
> `enrollmentID, affiliation, maxEnrollments, signingIdentity, enrollmentSecret, role, attrs, `

I don't think we should be making any breaking changes to existing functions in the public API outside of a major release version. Once we've done a v2.2.0 release alongside the Fabric v2.2 LTS release, perhaps we should cut a v2 branch so we don't have to be so careful about breaking changes in master branch. We still do need to keep in mind that any breaking changes we make do impact on client applications using the SDK, so they aren't something we should do lightly.

davidkhala commented:
> We still do need to keep in mind that any breaking changes we make do impact on client applications using the SDK, so they aren't something we should do lightly.

Agree on most of points. But my assumption is fabricCaClient.js is not directly exposed to first-tier user as external. It stays behind fabricCaServices and do not have typescript definition for it.
